### DISCUSSION TITLE
A question about the kernel

### DISCUSSION THREAD

**Discussion by wowodejiajia** [idaholab/moose#22011] [Q&A General] [answered]
Dear MOOSE experts,

In the original work, the author compiled the program himself. Do you know whether there are ready-made Kernels to choose from? I have excerpted some of the content.

```
[Variables]
  # variables. w: chemical potential, eta: order parameter, pot: applied overpotential.  
  [./w]
  [../]
  [./eta]
  [../]
  [./pot]
  [../]
[]

[Materials]
  [./Bultervolmer]
    type = DerivativeParsedMaterial
    function = 'Ls*(exp(pot*AA/2.)+14.89*cl*(1-h)*exp(-pot*AA/2.))*dh'
    args = 'pot eta w'
    f_name = G
    derivative_order = 1
    material_property_names = 'Ls dh:=D[h,eta] h cl:=D[f1,w] AA'
    outputs = exodus
  [../]
```

The author has compiled a kernel himself. Weak form is G*test.
```
[Kernels]
  [./BV]
    type = Kinetics
    variable = eta
    f_name = G
    cp=pot
    cv=eta
  [../]
```
Can "Kinetics" be replaced with a ready-made kernel?

I look forward to your reply.
Thank you!
[Kinetics.txt](https://github.com/idaholab/moose/files/9494636/Kinetics.txt)

**Comment by GiudGiud** [comment, marked as answer]
Hello

Since you have the source code, why dont you want to use Kinetics? 
You ll just need to update it to make sure the parameters are properly handled, see this post:
https://mooseframework.inl.gov/newsletter/2020_04.html#!

Should this post be categorized as phase field instead of general? What application is this for?

Guillaume

**Reply by wowodejiajia** [reply]
The first term of this equation is the phase field equation. The second term is the electrochemical kinetic equation.
![eq1](https://user-images.githubusercontent.com/85658962/188782767-b93eb02e-6f98-4a2e-99a8-1eb6589c5631.png)

You're right. Kinetics can operate normally after being updated. However, I want to know if there is a ready-made kernel like this F(a,b,c)=0 (a, b and c are variables. F is a complex equation defined in [Materials]).

**Reply by GiudGiud** [reply]
so you want to be solving an equation to compute F on each quadrature point?
How difficult is this equation? The Kinetics kernel doesnt have this, F and its derivatives are computed elsewhere

**Reply by wowodejiajia** [reply]
What I said before may not be accurate. Let me put it another way. 
Is there a kernel similar to Kinetics kernel？I looked on the website and only found BodyForce kernel. Used in this program
[./BV]
    type = BodyForce
    variable = eta
    function = G   # It seems wrong here
[../]

**Reply by GiudGiud** [reply]
BodyForce doesnt have the derivative term you have in Kinetics. If you work with AD, you may use ADBodyForce to get the derivatives right
 
imo you should just use kinetics if you know it works for you

**Reply by wowodejiajia** [reply]
Oh, I see. Thank you.
